### PR TITLE
Parse header md level

### PR DESCRIPTION
**Pull request by CorneliaKelinske**
I have simplified some stuff and added comments in the code regarding what I did.

I feel that in order to simplify even more, I would have to change the whole process flow of the original code.
It seems to me that in the original version the code is checked for headers, list, paras first and then for highlighting.
If I was to set this up, I would probably do the highlighting first and then process the rest.

What are your thoughts?
How much else do you think I should do?
where do I start?

### CODE REVIEW

File: elixir/markdown/lib/markdown.ex [hunk 1]
```diff
@@ -11,37 +11,53 @@ defmodule Markdown do
...
   def parse(m) do
-    patch(Enum.join(Enum.map(String.split(m, "\n"), fn t -> process(t) end)))
-  end
+    m
+    |> String.split("\n")
+    |> IO.inspect(label: "output of String.split")
+    |> Enum.map(& process(&1))
+    |> IO.inspect(label: "Enum.map process output")
+    |> Enum.join()
+    |> IO.inspect(label: "output of Enum.join")
```
Comment (alanvardy) [inline elixir/markdown/lib/markdown.ex]: Do you want to remove the `IO.inspect`s? Using a tool like [Credo](https://github.com/rrrene/credo) can automate these checks for you.

File: elixir/markdown/lib/markdown.ex [hunk 2]
```diff
@@ -11,37 +11,53 @@ defmodule Markdown do
...
+    |> patch()
```
Comment (alanvardy) [inline elixir/markdown/lib/markdown.ex:21]: Nice work on the piping

File: elixir/markdown/lib/markdown.ex [hunk 3]
```diff
@@ -11,37 +11,53 @@ defmodule Markdown do
...
 
+  end
 
-  defp process(t) do
-    case String.first(t) do
-      "#" -> enclose_with_header_tag(parse_header_md_level(t))
-      "*" -> parse_list_md_level(t)
-      _ -> enclose_with_paragraph_tag(String.split(t))
+  #I removed the nested if structure and replaced it with 'case'"
+  # changed 't' into 'line' because it is the result of String.split in the parse function
```
Comment (alanvardy) [inline elixir/markdown/lib/markdown.ex:26]: :+1:

File: elixir/markdown/lib/markdown.ex [hunk 4]
```diff
@@ -11,37 +11,53 @@ defmodule Markdown do
     "<h1>Header!</h1><ul><li><em>Bold Item</em></li><li><i>Italic Item</i></li></ul>"
   """
 
-  #This should probably be a pipeline instead of all those nested functions
+  #Refactored into a pipeline to get rid of the nested functions
   @spec parse(String.t()) :: String.t()
   def parse(m) do
-    patch(Enum.join(Enum.map(String.split(m, "\n"), fn t -> process(t) end)))
-  end
+    m
```
Comment (alanvardy) [inline elixir/markdown/lib/markdown.ex]: Is there a better variable that could be used rather than `m`?

File: elixir/markdown/lib/markdown.ex [hunk 5]
```diff
@@ -11,37 +11,53 @@ defmodule Markdown do
...
   end
 
-  defp parse_header_md_level(hwt) do
-    [h | t] = String.split(hwt)
+  #decided to change the 'hwt' argument to 'line' since it is still the line from the process function that gets passed in here
+  defp parse_header_md_level(line) do
+    [h | t] = String.split(line)
     {to_string(String.length(h)), Enum.join(t, " ")}
+    |> IO.inspect(label: "output of parse_header_md_level")
+    # {"number", "text minus the header markup"}
```
Comment (alanvardy) [inline elixir/markdown/lib/markdown.ex]: Is this a forgotten comment?

One thing that might be able to improve the readability of line 41 is some well-named variables.

File: elixir/markdown/lib/markdown.ex [hunk 6]
```diff
@@ -11,37 +11,53 @@ defmodule Markdown do
...
+  defp parse_list_md_level(line) do
+    IO.inspect(line, label: "Input to parse_list_md_level")
+    text = String.split(String.trim_leading(line, "* "))
+    IO.inspect(text, label: "This is text")
+    "<li>" <> join_words_with_tags(text) <> "</li>"
+    |> IO.inspect(label: "parse_list_md_level output")
   end
 
   defp enclose_with_header_tag({hl, htl}) do
     "<h" <> hl <> ">" <> htl <> "</h" <> hl <> ">"
```
Comment (alanvardy) [inline elixir/markdown/lib/markdown.ex:48]: Is the htl variable just text? Can that variable name be improved?

**Review by alanvardy — COMMENTED** [review]
Looking good, the rampant `IO.inspects make it a bit harder to read and probably clutter up the terminal a bit?

Feel free to ping me for a review whenever :+1:

**Comment by CorneliaKelinske** [comment]
Let me know if I can submit this to exercism now. I'm kinda done effing around with it.
Effing is not a bad word

**Comment by alanvardy** [comment]
> Let me know if I can submit this to exercism now. I'm kinda done effing around with it.
> Effing is not a bad word

File: elixir/markdown/lib/markdown.ex (hunk 1 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex]: You can respond to the above comment or click "resolve conversation" to collapse this thread!

File: elixir/markdown/lib/markdown.ex (hunk 4 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex]: No response to this question yet

File: elixir/markdown/lib/markdown.ex (hunk 5 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex]: ?

File: elixir/markdown/lib/markdown.ex (hunk 6 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex:49]: ?

File: elixir/markdown/lib/markdown.ex (hunk 1 above)
Comment (CorneliaKelinske) [reply elixir/markdown/lib/markdown.ex]: There should be no more IO.inspects now

File: elixir/markdown/lib/markdown.ex (hunk 4 above)
Comment (CorneliaKelinske) [reply elixir/markdown/lib/markdown.ex]: I have changed it to "markdown_text". I'm not 100% sure if this is a good choice. Maybe 'source_text'?

File: elixir/markdown/lib/markdown.ex (hunk 5 above)
Comment (CorneliaKelinske) [reply elixir/markdown/lib/markdown.ex]: The forgotten comment is deleted and I have changed the variables.

**Comment by CorneliaKelinske** [comment]
Thank you for the in-depth review. I have addressed all the comments and will upload the code with the latest changes now.
Please note: I have also added comments throughout the code that explain the changes I have made.
Let me know if there is anything else that sould be changed befor I merge.
Thank you!

File: elixir/markdown/lib/markdown.ex (hunk 1 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex]: :+1:

File: elixir/markdown/lib/markdown.ex (hunk 4 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex]: I think either sounds good, maybe `markdown_text` could be better as it is a bit more specific?

File: elixir/markdown/lib/markdown.ex (hunk 5 above)
Comment (alanvardy) [reply elixir/markdown/lib/markdown.ex]: :+1:

**Review by alanvardy — APPROVED** [review]
Looks great! :star2:

**Pull request merged**